packages/pyright-internal/src/analyzer/typeEvaluator.ts

```diff
@@ -15091,7 +15091,11 @@ export function createTypeEvaluator(importLookup: ImportLookup, evaluatorOptions
             dataClassBehaviors = effectiveMetaclass.details.classDataClassTransform;
         } else {
             const baseClassDataTransform = classType.details.mro.find((mroClass) => {
-                return isClass(mroClass) && mroClass.details.classDataClassTransform !== undefined;
+                return (
+                    isClass(mroClass) &&
+                    mroClass.details.classDataClassTransform !== undefined &&
+                    mroClass !== classType
```

Comment:
A direct comparison may not work here for generic classes. That's classes in the `mro` array are specialized (i.e. `ClassType.cloneForSpecialization` is called). For example, if `Foo[A]` derives from `list[B]` which derives from `MutableSequence[C]`, then the mro list will include `Foo[A]`, `list[A]`, and `MutableSequence[A]`. 

Rather than using a simple `!==` check here, it is better to use `!ClassType.isSameGenericClass(mroClass, classType)`. This will ignore any differences in specialization.